File: targets/java/micronaut/src/main/java/rb/micronaut/routes/Cached.java
package rb.micronaut.routes;

import io.micronaut.http.HttpHeaders;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Header;
import rb.domain.Domain;

/**
 * cached: validator headers and the conditional request.
 *
 * The ETag is pinned in the fixture, so this measures emitting the header and comparing it
 * rather than hashing the body.
 *
 * The comparison requires a non-empty header: matching a missing if-none-match against an
 * empty ETag answers 304 to a client that never asked a conditional question.
 */
@Controller
public class Cached {

  private static HttpResponse<?> serve(String size, String inm) {
    String etag = Domain.etagOf(size);
    if (!inm.isEmpty() && inm.equals(etag)) {
      return HttpResponse.status(HttpStatus.NOT_MODIFIED)
                         .header(HttpHeaders.ETAG, etag)
                         .header(HttpHeaders.CACHE_CONTROL, Domain.CACHEABLE)
                         .header("x-rb-serial", Domain.nextSerial());
    }
    return HttpResponse.ok(Domain.payload(size))
                       .header(HttpHeaders.ETAG, etag)
                       .header(HttpHeaders.CACHE_CONTROL, Domain.CACHEABLE)
                       .header("x-rb-serial", Domain.nextSerial());
  }

  @Get("/cached/small")
  HttpResponse<?> small(@Header(name = "if-none-match", defaultValue = "") String inm) {
    return serve("small", inm);
  }

  @Get("/cached/medium")
  HttpResponse<?> medium(@Header(name = "if-none-match", defaultValue = "") String inm) {
    return serve("medium", inm);
  }

  @Get("/cached/large")
  HttpResponse<?> large(@Header(name = "if-none-match", defaultValue = "") String inm) {
    return serve("large", inm);
  }
}

File: targets/dotnet/minimal-apis/Routes/Cached.cs
using RequestBench.Domain;

namespace RequestBench.MinimalApis.Routes;

/// <summary>
/// cached: validator headers and the conditional request.
///
/// The ETag is pinned in the fixture, so this measures emitting the header and comparing it
/// rather than hashing the body. The size is closed over rather than read back out of the
/// path, and the comparison requires a non-empty header: matching a missing if-none-match
/// against an empty ETag answers 304 to a client that never asked a conditional question.
/// </summary>
public static class Cached
{
    private static Func<HttpContext, DomainModel, IResult> Serve(string size) =>
        (context, model) =>
        {
            string etag = model.ETagOf(size);
            HttpResponse response = context.Response;
            response.Headers.ETag = etag;
            response.Headers.CacheControl = DomainModel.Cacheable;
            response.Headers["x-rb-serial"] = model.NextSerial();
            string inm = context.Request.Headers.IfNoneMatch.ToString();
            return inm.Length > 0 && inm == etag
                ? Results.StatusCode(304)
                : Results.Json(model.Payload(size));
        };

    public static void Map(WebApplication app)
    {
        app.MapGet("/cached/small", Serve("small"));

        app.MapGet("/cached/medium", Serve("medium"));

        app.MapGet("/cached/large", Serve("large"));
    }
}
